platform-sdk/docs/proposals/healthMonitor/healthMonitorProposal.md

```diff
@@ -0,0 +1,127 @@
+- The backpressure mechanism should not be overly aggressive. That is, we shouldn't see
+  it engaging under normal circumstances on nodes that are not in a state of distress.
+- The implementation of the backpressure mechanism should be sufficiently clean and
+  maintainable. A performant solution that is too complex is not a solution we should
+  accept as a solution.
+
+### Design Decisions
+
+This change can be split into two distinct sub-problems: the detection of a system in duress, and the action that
+is taken to apply backpressure on a distressed system.
+
+#### Detection
+
+The current platform is built from a number of "components". A component is usually a unit of logic that runs on its
+own thread (or logical thread). Each component has a queue (or logical queue) of data it is ingesting. If a component
+produces output, that output is enqueued within the queues of other components.
+
+For each component with a queue, define a preferred size. The queue will not apply hard backpressure even if the
+size of the queue exceeds the preferred size. That is, insertion into the queue will never block. However, a queue
+whose size exceeds its preferred size will be considered to be "unhealthy".
+
+In the background, a new component called the "health monitor" will periodically check the health of each queue.
+Checking the health of a queue is cheap (i.e. an atomic long read). The frequency of polling is configurable, with the
+default being 10hz.
+
+The health monitor records the length of time that each queue has been in an unhealthy state. If a queue has been
+unhealthy for one minute, that is a much stronger signal than if a queue has been unhealthy for a few hundred
+milliseconds. If a previously unhealthy queue is observed as being healthy, the timer for that queue is immediately
+reset to 0.
+
+When the health monitor reports the total health of the system, it reports that value as a duration. Specifically,
+it reports the amount of time that the most unhealthy queue has been unhealthy. If there is one queue that has been
+unhealthy for 10s and another that has been unhealthy for 100s, it will report that the system unhealthiness factor
+is 100s. If the second queue clears its backlog and becomes healthy, it will then report that the system unhealthiness
+factor is 10s.
+
+The health monitor publishes health reports over a wire. If it sends a duration of 0s, that means that the system is in
+a healthy state. As the system becomes more and more unhealthy, it will periodically publish higher and higher
+durations.
+
+#### Reaction
+
+Various subsystems react to the health monitor's reports in different ways.
+
+##### Event Creation
+
+If the health monitor reports an unhealthy duration exceeding X seconds (configurable), then the creation of self
+events is disabled. Once the system reports an unhealthy duration below X seconds, event creation may resume.
```

Comment:
The proposal looks good overall. However, there's quite a number of configurable parameters that can change the performance profile of this implementation significantly. 

They will need tuning initially. And they will need re-tuning when the code in hedera-services changes significantly enough to make the old parameter values inefficient, which I'm afraid may happen more often than we'd like.

Apart from this concern, the proposal looks good to me.

Reply:
+ I would like to see an overview of all needed config params (including a short description, param type and in best case default value)